core/src/main/scala/kafka/admin/TopicCommand.scala

```diff
@@ -17,248 +17,403 @@
-        adminZkClient.addPartitions(topic, existingAssignment, allBrokers, nPartitions, newAssignment)
-        println("Adding partitions succeeded!")
       }
     }
-  }
 
-  def listTopics(zkClient: KafkaZkClient, opts: TopicCommandOptions) {
-    val topics = getTopics(zkClient, opts)
-    for(topic <- topics) {
-      if (zkClient.isTopicMarkedForDeletion(topic)) {
-        println("%s - marked for deletion".format(topic))
+    override def deleteTopic(opts: TopicCommandOptions): Unit = {
+      val topics = getTopics(opts.topic, opts.excludeInternalTopics)
+      if (topics.isEmpty && !opts.ifExists) {
+        throw new IllegalArgumentException(s"Topic ${opts.topic} does not exist")
+      }
+      adminClient.deleteTopics(topics.asJavaCollection).all().get()
+    }
+
+    override def getTopics(topicWhitelist: Option[String], excludeInternalTopics: Boolean = false): Seq[String] = {
+      val allTopics = if (excludeInternalTopics) {
+        adminClient.listTopics()
       } else {
-        println(topic)
+        adminClient.listTopics(new ListTopicsOptions().listInternal(true))
       }
+      doGetTopics(allTopics.names().get().asScala.toSeq.sorted, topicWhitelist, excludeInternalTopics)
     }
+
+    override def close(): Unit = adminClient.close()
   }
 
-  def deleteTopic(zkClient: KafkaZkClient, opts: TopicCommandOptions) {
-    val topics = getTopics(zkClient, opts)
-    val ifExists = opts.options.has(opts.ifExistsOpt)
-    if (topics.isEmpty && !ifExists) {
-      throw new IllegalArgumentException("Topic %s does not exist on ZK path %s".format(opts.options.valueOf(opts.topicOpt),
-          opts.options.valueOf(opts.zkConnectOpt)))
+  class ZookeeperTopicService(zkClient: KafkaZkClient) extends TopicService {
+
+    def this(zkConnect: Option[String]) {
+      this(KafkaZkClient(zkConnect.get, JaasUtils.isZkSecurityEnabled, 30000, 30000,
+        Int.MaxValue, Time.SYSTEM))
     }
-    topics.foreach { topic =>
+
+    override def createTopic(topic: CommandTopicPartition): Unit = {
+      val adminZkClient = new AdminZkClient(zkClient)
       try {
-        if (Topic.isInternal(topic)) {
-          throw new AdminOperationException("Topic %s is a kafka internal topic and is not allowed to be marked for deletion.".format(topic))
-        } else {
-          zkClient.createDeleteTopicPath(topic)
-          println("Topic %s is marked for deletion.".format(topic))
-          println("Note: This will have no impact if delete.topic.enable is not set to true.")
+        if (topic.hasReplicaAssignment)
+          adminZkClient.createTopicWithAssignment(topic.name, topic.configsToAdd, topic.replicaAssignment.get)
+        else
+          adminZkClient.createTopic(topic.name, topic.partitions.get, topic.replicationFactor, topic.configsToAdd, topic.rackAwareMode)
+        println("Created topic \"%s\".".format(topic.name))
+      } catch  {
+        case e: TopicExistsException => if (!topic.ifTopicDoesntExist()) throw e
```

Comment:
That is a weird logic to handle a conditional operation. :)
So there is an `--if-not-exists` option in TopicCommand which would create the topic if it doesn't exist. Now the implementation will try to create the topic anyways but specifying this flag will suppress the exception. I didn't refactor this too much because it'll be deprecated and didn't want to change its behavior. In `AdminClientTopicService.createTopic` it is implemented properly.

Reply:
Also looking at the bigger picture I'm not 100% convinced that we'll need the mentioned flag because for more advanced logic people should use the AdminClient directly, so we might even deprecate it.